### ISSUE TITLE
Chat app that handles different document layouts

### ISSUE THREAD

**Issue by RubenVerborgh** [SolidLabResearch/Challenges#1]
## Pitch
- Mashlib currently has a [short chat](https://drive.verborgh.org/public/2022/short-chat/) and a [long chat](https://drive.verborgh.org/public/2022/long-chat/) which are stored in different ways: short chats store all messages [in one single document](https://drive.verborgh.org/public/2022/short-chat/index.ttl), whereas long chats [create one document per day](https://drive.verborgh.org/public/2022/long-chat/index.ttl).
- Apps currently need to be coded against one of those options.
- Let's write a component such that apps can read/write messages from any document layout—including future layouts.

## Desired solution
- The properties of a chat document layout should be described in RDF:
  - "create one document for all messages"
  - "create one document per day"
  - "store each message in the pod of the user who wrote that message"
  - …
 - This component should be able to understand such layouts, and read/write files accordingly.

## Acceptance criteria
- A browser UI component that accepts both short chats and long chats
- The logic for reading chats should be in a separate library/component that can be used by others

## Pointers
- https://ruben.verborgh.org/blog/2019/06/17/shaping-linked-data-apps/ (footprints)
- LDES / LDES Write

## Scenarios
_(none yet)_

**Comment by pietercolpaert** [comment]
A sub-challenge I’m interested in: _when_ would a Solid app choose to:
 1. write everything in one big file, 
 2. write everything in separate files but in the same container,
 3. write everything in X items per page?

Hypothesis: when using HTTP/2, the difference between 2 and 3 should be negligible.

**Comment by RubenVerborgh** [comment]
Hopefully the _app_ would never choose, as it should interact with a knowledge graph, not an interface. (But we're not there yet.)

**Comment by pietercolpaert** [comment]
Further thoughts: since mainly the Solid auth workflow is per resource, even HTTP/2 would be too slow, so we will try to find a way to bring multiple small resources together into one (“virtual containers”)

**Comment by Dexagod** [comment]
Has there been any update on new ontologies or something for chats in Solid?
Have these interactions been specified anywhere, which predicates are expected where?

**Comment by Dexagod** [comment]
We can make use of the created frontends for the VOP project here:
https://github.com/SolidLabResearch/SolidChatApp

and use this as a basis to create a backend layer for the challenge

**Comment by RubenVerborgh** [comment]
Here's an existing app that was built using shapes: https://forum.solidproject.org/t/photo-album-app/5260/7
Might be interesting as inspiration.

**Comment by pheyvaer** [comment]
@Dexagod @andreipopi What is the status of this challenge?

**Comment by andreipopi** [comment]
Dear Pieter,


Probably you have been informed that after the Gentse Feesten I was told by Pieter B. and Pieter C. that I will not stay in the group after October, and that I would be free to pursue my next endeavours.


This means that the meeting that was supposed to happen with Ruben D. in order to kick start that challenge never happened, and the challenge remained in the initial state, on which there seemed to already be some confusion even in the last meeting when I was informed that I should leave. In particular, Pieter B. was puzzled on why I looked into footprints which are clearly mentioned in the challenge itself (or at least were at that point) as being relevant, instead of somehow focusing on what I already did for the students’ chat.


In summary, my position in the group was cut, and the engineering for that challenge never started.


I wish you all the best with the challenges.


Kind regards,

Andrei Popescu

Inviato da iPhone

Il giorno 2 set 2022, alle ore 09:24, Pieter Heyvaert ***@***.***> ha scritto:

﻿

@Dexagod<https://github.com/Dexagod> @andreipopi<https://github.com/andreipopi> What is the status of this challenge?

**Comment by pheyvaer** [comment]
@Dexagod What is the status of this?

**Comment by pietercolpaert** [comment]
I just assigned @woutslabbinck  to do:
 1. An implementation using [LDES in LDP specification](https://woutslabbinck.github.io/LDESinLDP/) and his libraries he developed at https://www.npmjs.com/package/@treecg/versionawareldesinldp - according to this specification
 2. Identify one of the many open-source prototypes of a Solid Chat that could serve as a demo to get this into, in which you will be able to configure the parameters of how to write
 3. A separate library should be able to show a chat regardless of file-structure it is stored in

**Comment by pheyvaer** [comment]
@woutslabbinck Are you still working on this? If so, what is the status of this?

**Comment by woutslabbinck** [comment]
Since my assignment, I have not found time yet to execute the plan that Pieter Colpaert has proposed.

**Comment by pheyvaer** [comment]
@woutslabbinck Is there any update on this?

**Comment by woutslabbinck** [comment]
I am afraid it's the same answer as last time. I have not found time yet to focus on this challenge.

**Comment by pheyvaer** [comment]
Ok, then I'm gonna assign you to make clear that nobody is working on it at the moment.
